core-primitives/stf-state-handler/src/state_snapshot_repository_loader.rs

```diff
@@ -0,0 +1,202 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::{
+	error::Result,
+	file_io::StateFileIo,
+	state_snapshot_primitives::{
+		initialize_shard_with_snapshot, SnapshotHistory, StateId, StateSnapshotMetaData,
+	},
+	state_snapshot_repository::StateSnapshotRepository,
+};
+use itp_types::ShardIdentifier;
+use log::*;
+use std::{
+	collections::VecDeque, fmt::Debug, iter::FromIterator, marker::PhantomData, sync::Arc, vec::Vec,
+};
+
+/// Loads a state snapshot repository from existing shards directory with state files.
+pub struct StateSnapshotRepositoryLoader<FileIo, State, HashType> {
+	file_io: Arc<FileIo>,
+	phantom_data: PhantomData<(State, HashType)>,
+}
+
+impl<FileIo, State, HashType> StateSnapshotRepositoryLoader<FileIo, State, HashType>
+where
+	FileIo: StateFileIo<StateType = State, HashType = HashType>,
+	HashType: Copy + Eq + Debug,
+{
+	pub fn new(file_io: Arc<FileIo>) -> Self {
+		StateSnapshotRepositoryLoader { file_io, phantom_data: Default::default() }
+	}
+
+	/// Load a state snapshot repository from an existing set of files and directories.
+	pub fn load_snapshot_repository(
+		&self,
+		snapshot_history_cache_size: usize,
+	) -> Result<StateSnapshotRepository<FileIo, State, HashType>> {
+		let snapshot_history = self.load_state_snapshot_history()?;
+
+		StateSnapshotRepository::new(
+			self.file_io.clone(),
+			snapshot_history_cache_size,
+			snapshot_history,
+		)
+	}
+
+	fn load_state_snapshot_history(&self) -> Result<SnapshotHistory<HashType>> {
```

Comment:
```suggestion
	fn load_and_initialize_state_snapshot_history(&self) -> Result<SnapshotHistory<HashType>> {
```

Reply:
Here the `initialize` is correct, I can add it 👍